### PR TITLE
[android] Improve route elevation profile and fix RTL layout

### PR DESCRIPTION
**Pull request by strider-dunadan**
## Summary                                                                                                                                               
  - Separate route elevation profile layout from track elevation profile, removing unused `FloatingMarkerView`                                             
  - Fix ascent/descent text rendering in RTL locales using `BidiFormatter`
  - Clear chart data and elevation marker when hiding altitude panel or switching routes                                                                   
  - Simplify `restoreRoutingPanelState` by removing redundant `post()` deferral                                                                            
  - Remove redundant `isVehicleRouterType` guard in `showRouteAltitudeChart()` (already checked by caller)
  - Fix interpolation epsilon in `RoutingManager::GetRoutePointAtDistance` and add unit tests

### CODE REVIEW

File: libs/map/routing_manager.cpp [hunk 1]
```diff
@@ -1253,9 +1253,9 @@ std::optional<m2::PointD> RoutingManager::GetRoutePointAtDistance(double distanc
 
   auto const it = std::upper_bound(distances.begin(), distances.end(), distanceMeters);
   size_t const idx = std::distance(distances.begin(), it);
-  // distances[idx-1] < distanceMeters <= distances[idx], points[idx] .. points[idx+1]
+  // distances[idx-1] <= distanceMeters < distances[idx], points[idx] .. points[idx+1]
   double const segLen = distances[idx] - (idx > 0 ? distances[idx - 1] : 0.0);
-  double const f = segLen > 1e-9 ? (distanceMeters - (idx > 0 ? distances[idx - 1] : 0.0)) / segLen : 0.0;
+  double const f = segLen > 0 ? (distanceMeters - (idx > 0 ? distances[idx - 1] : 0.0)) / segLen : 0.0;
```
Comment (biodranik) [inline libs/map/routing_manager.cpp]: The comment correction on line 1256 is good — `upper_bound` gives `distances[idx-1] <= distanceMeters < distances[idx]`, so the half-open interval is now right.

The `segLen > 1e-9` → `segLen > 0` change, however, doesn't fix a real bug and the commit message's "Fix interpolation epsilon" framing is misleading. Given the preconditions (`<= 0` early return, `>= distances.back()` early return), `upper_bound` cannot land at an index where `distances[idx] == distances[idx-1]`: if distances contain a run of duplicates, upper_bound skips past all of them to the first strictly-greater element, and the resulting `segLen` is always the gap to the next unique value. I couldn't construct any `(distances, points, distanceMeters)` triple that reaches this line with `segLen == 0`.

Please either:
- drop the epsilon change and keep only the comment fix, or
- keep `segLen > 0` as defensive hardening but reword the commit message to say so, and replace the silent `f = 0.0` fallback with `CHECK(segLen > 0, (distanceMeters, distances[idx-1], distances[idx]))` — an unreachable branch shouldn't masquerade as logic.

File: libs/map/map_tests/track_tests.cpp [hunk 2]
```diff
@@ -13,13 +13,36 @@
...
+
+  if (distanceMeters >= distances.back())
+    return points.back();
+
+  auto const it = std::upper_bound(distances.begin(), distances.end(), distanceMeters);
+  size_t const idx = std::distance(distances.begin(), it);
+  double const segLen = distances[idx] - (idx > 0 ? distances[idx - 1] : 0.0);
+  double const f = segLen > 0 ? (distanceMeters - (idx > 0 ? distances[idx - 1] : 0.0)) / segLen : 0.0;
+  return points[idx] + (points[idx + 1] - points[idx]) * f;
+}
```
Comment (biodranik) [inline libs/map/map_tests/track_tests.cpp]: `InterpolateByDistance` is a copy-paste of the body of `RoutingManager::GetRoutePointAtDistance`. The tests therefore exercise this local clone, not the real production code — if someone refactors `GetRoutePointAtDistance` tomorrow, these tests pass regardless.

Extract the pure logic into a free helper (e.g. `routing::InterpolatePointAtDistance(distances, points, meters)` in `routing_manager.hpp/.cpp`) and have both `GetRoutePointAtDistance` and the tests call it. That way the test coverage is load-bearing.

File: libs/map/map_tests/track_tests.cpp [hunk 3]
```diff
@@ -194,6 +217,43 @@ UNIT_TEST(Track_GetPoint_NegativeDistance)
...
+
+UNIT_TEST(InterpolateByDistance_ZeroLengthSegment)
+{
+  std::vector<m2::PointD> const points = {{0, 0}, {0, 0}, {1, 0}};
+  std::vector<double> const distances = {0.0, 1.0};
+
+  auto const result = InterpolateByDistance(distances, points, 0.5);
+  TEST(!std::isnan(result.x), ());
+  TEST(!std::isnan(result.y), ());
+}
```
Comment (biodranik) [inline libs/map/map_tests/track_tests.cpp:236]: This test doesn't actually exercise a zero-length segment. With `distances = {0.0, 1.0}` and `distanceMeters = 0.5`:
- `upper_bound(0.5)` returns the iterator at `distances[1] = 1.0` → `idx = 1`
- `segLen = distances[1] - distances[0] = 1.0 - 0.0 = 1.0` (not zero)
- `f = 0.5`, result = `points[1] + (points[2] - points[1]) * 0.5 = (0.5, 0)`

The duplicated points in `points` don't affect interpolation — the algorithm works on cumulative distances, not on pairwise Cartesian gaps. As a result this test only asserts "well-behaved input stays well-behaved" and the `TEST(!std::isnan(...))` checks never meaningfully run.

Given the analysis in the routing_manager.cpp comment, `segLen == 0` is actually unreachable under the current preconditions — so please either delete this test outright, or rename it to something that reflects what it really checks (e.g. `InterpolateByDistance_DuplicatedPoints`) and add a brief comment explaining it's a smoke test, not zero-length coverage.

File: android/app/src/main/java/app/organicmaps/routing/RoutingBottomMenuController.java [hunk 4]
```diff
@@ -362,27 +357,24 @@ void restoreRoutingPanelState(@NonNull Bundle state)
...
     UiUtils.show(mAltitudeChart);
 
-    mAltitudeDifference.setText(mContext.getString(R.string.route_ascent_descent_format,
-                                                   Framework.nativeFormatAltitude(data.getTotalAscent()),
-                                                   Framework.nativeFormatAltitude(data.getTotalDescent())));
+    final BidiFormatter bidi = BidiFormatter.getInstance();
+    final String ascentDescent = mContext.getString(
+        R.string.route_ascent_descent_format, bidi.unicodeWrap(Framework.nativeFormatAltitude(data.getTotalAscent())),
+        bidi.unicodeWrap(Framework.nativeFormatAltitude(data.getTotalDescent())));
+    mAltitudeDifference.setText(ascentDescent);
```
Comment (biodranik) [inline android/app/src/main/java/app/organicmaps/routing/RoutingBottomMenuController.java]: `BidiFormatter.unicodeWrap` on the `%1$s`/`%2$s` substitutions protects each formatted number from being reordered internally, but the overall format `↗\u00A0%1$s ↘\u00A0%2$s` is still laid out by the TextView's bidi algorithm. In an Arabic/Hebrew locale the entire run will be rendered right-to-left and the user sees `m 88 ↘ m 100 ↗` visually — inverting the semantic meaning of the ↗/↘ arrows.

Fix: wrap the full result and force LTR regardless of context, e.g.:

```java
final BidiFormatter bidi = BidiFormatter.getInstance(false /* rtlContext */);
final String ascent  = bidi.unicodeWrap(Framework.nativeFormatAltitude(data.getTotalAscent()));
final String descent = bidi.unicodeWrap(Framework.nativeFormatAltitude(data.getTotalDescent()));
final String raw = mContext.getString(R.string.route_ascent_descent_format, ascent, descent);
mAltitudeDifference.setText(bidi.unicodeWrap(raw, TextDirectionHeuristics.LTR));
```

The PR summary says this fixes RTL rendering — please verify by actually switching the device to Arabic/Hebrew (Settings → System → Languages) and attach a screenshot in the PR description. As currently written I don't think the observable result is correct.

File: android/app/src/main/java/app/organicmaps/routing/RoutingBottomMenuController.java [hunk 5]
```diff
@@ -498,6 +490,7 @@ else if (id == R.id.btn__manage_route)
       mListener.onManageRouteOpen();
     else if (id == R.id.btn__save)
     {
+      // Clear elevation preview marker before saving.
       Framework.nativeRouteRemoveElevationActivePoint();
       Framework.nativeSaveRoute();
```
Comment (biodranik) [inline android/app/src/main/java/app/organicmaps/routing/RoutingBottomMenuController.java:500]: This clears the native drape selection, but it does not clear MPAndroidChart's internal highlight in `RouteElevationChartController.mChart`.

Repro: (1) build a pedestrian/bicycle route, (2) scrub the elevation chart so a preview marker appears on the map, (3) tap Save. After step 3 the map marker is dismissed (good) but the chart still draws the highlight circle at the scrubbed distance — chart and map are out of sync until the user taps the chart again or the panel is rebuilt. Since this PR's explicit intent (see the new comment on line 493) is to reset the preview state before saving, the chart side needs the same treatment.

Suggested fix — add a public helper to `RouteElevationChartController`:

```java
public void clearSelection()
{
  if (mChart != null)
    mChart.highlightValue(null, false);  // false: don't call listener; native marker is cleared explicitly below
}
```

and call it from this save handler *before* the native clear:

```java
mRouteElevationChartController.clearSelection();
Framework.nativeRouteRemoveElevationActivePoint();
Framework.nativeSaveRoute();
```

Alternatively, call `mChart.highlightValue(null, true)` and let the existing `onNothingSelected` listener handle the native side, removing the explicit `nativeRouteRemoveElevationActivePoint()` call here.

File: android/app/src/main/java/app/organicmaps/routing/RoutingBottomMenuController.java [hunk 6]
```diff
@@ -204,6 +205,7 @@ void showAltitudeChartAndRoutingDetails()
 
   void hideAltitudeChartAndRoutingDetails()
   {
+    mRouteElevationChartController.setData(null);
```
Comment (biodranik) [inline android/app/src/main/java/app/organicmaps/routing/RoutingBottomMenuController.java:210]: `mRouteElevationChartController.setData(null)` cascades through `RouteElevationChartController.setData` → `mListener.onElevationPointDeselected()` → `Framework.nativeRouteRemoveElevationActivePoint()`. That's correct when hiding, but it also fires on line 365 (empty-data early return) and every time `setData(data)` is called with a new non-null route (line 370 → controller lines 102-103 always deselect before applying fresh data). On a route build there's no marker to clear, so the JNI → `DeselectObject` round-trip is wasted work on the hot path.

Please split the concerns: make `setData(...)` purely set chart data, and add an explicit `clearSelection()` (see the save-button comment on line 495). Call `clearSelection()` from `hideAltitudeChartAndRoutingDetails()` and the save handler, and drop the `onElevationPointDeselected()` calls from `setData`. That also removes the risk of a double-call when `mChart.clear()` itself triggers `onNothingSelected()`.

File: android/app/src/main/res/layout/route_elevation_profile.xml [hunk 7]
```diff
@@ -0,0 +1,47 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="@dimen/elevation_profile_content_height"
+    android:background="?cardBackground"
+    android:orientation="vertical">
```
Comment (biodranik) [inline android/app/src/main/res/layout/route_elevation_profile.xml]: A few layout-hygiene items in this new file (the old `elevation_profile.xml` had the same issues but since you're creating a fresh file, now is the right time to clean them up):

1. **Indentation**: 4-space indent here, 2-space everywhere else in `res/layout/` (including the sibling `elevation_profile.xml` this was derived from). Please match the project convention so the diff stays clean.
2. **Line 6** — `android:orientation="vertical"` is silently ignored on `RelativeLayout` (not a `LinearLayout` attribute). Drop it.
3. **Lines 12-13** — `layout_alignParentStart="true"` + `layout_alignParentEnd="true"` on a `match_parent`-width child are no-ops. Drop them.
4. **Line 37** — `android:layout_gravity="center_horizontal|start"` is ignored on a direct child of `RelativeLayout` (RelativeLayout uses `layout_alignParent*`, not `layout_gravity`). Drop it.

File: android/app/src/main/java/app/organicmaps/maplayer/MapButtonsController.java [hunk 8]
```diff
@@ -56,6 +56,7 @@ public class MapButtonsController extends Fragment
   private LayersButton mToggleMapLayerButton;
   @Nullable
   FloatingActionButton mTrackRecordingStatusButton;
+
```
Comment (biodranik) [inline android/app/src/main/java/app/organicmaps/maplayer/MapButtonsController.java]: Unrelated whitespace change — this blank line has nothing to do with the route elevation profile or RTL fix, and the surrounding field declarations don't use blank-line grouping. Please revert.

File: android/app/src/main/res/layout-h400dp-land/map_buttons_layout_regular.xml [hunk 9]
```diff
@@ -24,7 +24,6 @@
       android:layout_alignParentBottom="true"
       app:layout_constraintStart_toStartOf="parent"
       app:layout_constraintTop_toTopOf="parent" />
-
   </androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (biodranik) [inline android/app/src/main/res/layout-h400dp-land/map_buttons_layout_regular.xml:27]: This hunk (and the identical trailing-blank-line cleanups in `layout-h400dp/map_buttons_layout_regular.xml` and `layout-land/map_buttons_layout_regular.xml`) has nothing to do with the route elevation profile / RTL fix that this PR is about. It's also internally inconsistent — `layout/map_buttons_layout_regular.xml` and the `map_buttons_layout_{navigation,planning}.xml` siblings still have their trailing blank lines.

Please either revert all three, or pull this out into a separate `[android] Clean up map button layouts whitespace` commit that touches all 11 `map_buttons_layout_*.xml` files consistently.

File: android/sdk/src/main/java/app/organicmaps/sdk/routing/RoutingController.java [hunk 10]
```diff
@@ -360,6 +360,7 @@ public void start()
     // and then app crashes. So, the previous route will be restored on the next app launch.
     saveRoute();
 
+    // Clear stale elevation preview marker.
```
Comment (biodranik) [inline android/sdk/src/main/java/app/organicmaps/sdk/routing/RoutingController.java:363]: Per `CLAUDE.md`: *"Default to writing no comments. Only add one when the WHY is non-obvious."* This comment just restates the method name — `nativeRouteRemoveElevationActivePoint()` is already self-explanatory. The interesting context (why we need to clear it *here* specifically — i.e. the user may have been scrubbing the elevation chart mid-planning, and we want to flush that preview state before transitioning to NAVIGATION) isn't in the comment.

Either drop this comment (and the one at line 459 — same nit — and the one at `RoutingBottomMenuController.java:493`) or rewrite them to carry the actual *why*.

File: android/app/src/main/res/layout-land/altitude_chart_panel.xml [hunk 11]
```diff
@@ -46,7 +46,8 @@
           tools:visibility="visible" />
     </LinearLayout>
 
-    <include layout="@layout/elevation_profile"
+    <include
+        layout="@layout/route_elevation_profile"
```
Comment (biodranik) [inline android/app/src/main/res/layout-land/altitude_chart_panel.xml]: The `layout="@layout/route_elevation_profile"` attribute was pulled onto its own line as part of this rename. Same gratuitous reformatting in `layout/altitude_chart_panel.xml:59-61` and `layout-w1020dp-land/altitude_chart_panel.xml:78-79`. It makes the diff noisier than necessary for what should be a 1-character change per file — please restore the original one-line form so the actual change is just the include target.

File: android/sdk/src/main/java/app/organicmaps/sdk/routing/RoutingController.java (hunk 10 above)
Comment (strider-dunadan) [reply android/sdk/src/main/java/app/organicmaps/sdk/routing/RoutingController.java:363]: Should be here

File: android/app/src/main/res/layout-land/altitude_chart_panel.xml (hunk 11 above)
Comment (strider-dunadan) [reply android/app/src/main/res/layout-land/altitude_chart_panel.xml]: Auto format

Comment (biodranik) [reply android/app/src/main/res/layout-land/altitude_chart_panel.xml]: We need to keep the same normalized formatting for all XML files. Any ideas on how to do that properly? 

This particular change looks unnecessary.

File: android/app/src/main/res/layout-w1020dp-land/altitude_chart_panel.xml [hunk 12]
```diff
@@ -75,7 +75,8 @@
           android:textSize="@dimen/text_size_routing_plan_detail_arrival"/>
     </LinearLayout>
 
-    <include layout="@layout/elevation_profile"
+    <include
```
Comment (biodranik) [inline android/app/src/main/res/layout-w1020dp-land/altitude_chart_panel.xml:78]: Unnecessary formatting here and in other places? It makes it harder to blame the git history and see why/when something was changed.

File: android/app/src/main/res/layout/elevation_profile.xml [hunk 13]
```diff
@@ -3,14 +3,11 @@
   xmlns:android="http://schemas.android.com/apk/res/android"
   android:layout_width="match_parent"
   android:layout_height="@dimen/elevation_profile_content_height"
-  android:background="?cardBackground"
-  android:orientation="vertical">
+    android:background="?cardBackground">
```
Comment (biodranik) [inline android/app/src/main/res/layout/elevation_profile.xml]: Ditto here and below. Please keep the formatting the same as the surrounding code and avoid unnecessary/unrelated changes.

File: android/app/src/main/res/layout/route_elevation_profile.xml [hunk 14]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
```
Comment (biodranik) [inline android/app/src/main/res/layout/route_elevation_profile.xml]: 2 spaces in other files, and 4 spaces here?

File: libs/map/routing_manager.cpp [hunk 15]
```diff
@@ -1235,15 +1235,10 @@ bool RoutingManager::GetRouteElevationInfo(ElevationInfo & ei) const
...
+                                      double distanceMeters)
 {
-  auto const * route = m_routingSession.GetRoute();
-  if (!route || !route->IsValid())
-    return std::nullopt;
-
-  auto const & distances = route->GetSegDistanceMeters();
-  auto const & points = route->GetPoly().GetPoints();
-  ASSERT_EQUAL(distances.size() + 1, points.size(), ());
+  CHECK_EQUAL(distances.size() + 1, points.size(), ());
```
Comment (vng) [inline libs/map/routing_manager.cpp]: Keep ASSERT here.

File: libs/map/routing_manager.cpp [hunk 16]
```diff
@@ -1253,12 +1248,26 @@ std::optional<m2::PointD> RoutingManager::GetRoutePointAtDistance(double distanc
...
   auto const it = std::upper_bound(distances.begin(), distances.end(), distanceMeters);
   size_t const idx = std::distance(distances.begin(), it);
-  // distances[idx-1] < distanceMeters <= distances[idx], points[idx] .. points[idx+1]
-  double const segLen = distances[idx] - (idx > 0 ? distances[idx - 1] : 0.0);
-  double const f = segLen > 1e-9 ? (distanceMeters - (idx > 0 ? distances[idx - 1] : 0.0)) / segLen : 0.0;
+  // distances[idx-1] <= distanceMeters < distances[idx], points[idx] .. points[idx+1]
+  double const prevDist = idx > 0 ? distances[idx - 1] : 0.0;
+  double const segLen = distances[idx] - prevDist;
+  // segLen == 0 is unreachable: upper_bound skips duplicates to the first strictly-greater element.
+  CHECK(segLen > 0, (distanceMeters, prevDist, distances[idx]));
```
Comment (vng) [inline libs/map/routing_manager.cpp]: Keep ASSERT here.

**Review by vng — APPROVED** [review]

**Comment by vng** [comment]
My point is: put assert where constraint is not critical e.g. we won't crash if it is not accomplished, but calculate some bad coordinates.
